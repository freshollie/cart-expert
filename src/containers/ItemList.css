.item-list {
    display: flex;
    flex-direction: row;
    flex-wrap: wrap;
    height: min-content;
}

.item-list > * {
    width: 150px;
    margin: 10px;
    padding: 10px;
    background-color: white;
    color: black;
    height: min-content;
    border-radius: 5px;
}