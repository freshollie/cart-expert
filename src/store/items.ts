import {
  createSelector,
  createSlice,
} from "@reduxjs/toolkit";
import { RootState } from ".";
import { Item } from "../types";

const ITEMS: readonly Item[] = [
  {
    id: "34583",
    name: "Beans",
    price: {
      amount: 50,
      divisor: 100,
    },
  },
  {
    id: "39845",
    name: "Coke",
    price: {
      amount: 70,
      divisor: 100,
    },
  },
  {
    id: "32844",
    name: "Oranges",
    price: {
      amount: 199,
      divisor: 100,
      perKg: true,
    },
  },
  {
    id: "34242",
    name: "Bananas",
    price: {
      amount: 199,
      divisor: 100,
      perKg: true,
    },
  },
];

export default createSlice({
  name: "items",
  initialState: ITEMS,
  reducers: {},
});

export const selectItemsLookup = createSelector(
  (state: RootState) => state.items,
  (items) =>
    items.reduce(
      (map, item) => ({ ...map, [item.id]: item }),
      {} as Record<string, Item>
    )
);

export const selectAllItems = (state: RootState) => state.items;

export const selectItembyId = createSelector(
  selectItemsLookup,
  (_: RootState, { id }: { id: string }) => id,
  (map, id) => map[id]
);
