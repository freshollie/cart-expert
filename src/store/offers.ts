import { createSelector, createSlice } from "@reduxjs/toolkit";
import { create } from "domain";
import { RootState } from ".";
import { Offer } from "../types";

// Offers are in a list as there could be more than 1 discount to apply on an item
const OFFERS: readonly Offer[] = [
  {
    itemId: "34583",
    qualifier: 3,
    name: "3 for 2 on all Beans",
    for: 2,
  },
  {
    itemId: "39845",
    qualifier: 2,
    name: "2 for Â£1 on all coke",
    price: {
      amount: 100,
      divisor: 100,
    },
  },
];

export default createSlice({
  initialState: OFFERS,
  name: "offers",
  reducers: {},
});

export const selectOffers = (state: RootState) => state.offers;
