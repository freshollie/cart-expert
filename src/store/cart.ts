import { createSelector, createSlice, PayloadAction } from "@reduxjs/toolkit";
import { RootState } from ".";
import { selectItemsLookup } from "./items";

const initialState = {} as Record<string, number>;
export default createSlice({
  name: "cart",
  initialState,
  reducers: {
    add: (
      state,
      { payload: { id, amount = 1 } }: PayloadAction<{ id: string; amount?: number }>
    ) => ({
      ...state,
      [id]: (state[id] ?? 0) + amount,
    }),
    remove: (
      state,
      { payload: { id, amount = 1 } }: PayloadAction<{ id: string; amount?: number }>
    ) => ({
      ...state,
      [id]: (state[id] ?? 0) - amount > 0 ? state[id]! - amount : 0,
    }),
    reset: () => initialState,
  },
});

export const selectCartAmount = (itemId: string) => (state: RootState) =>
  state.cart[itemId] ?? 0;

export const selectCartItems = createSelector(
  [(state: RootState) => state.cart, selectItemsLookup],
  (cart, items) =>
    Object.entries(cart)
      .map(([id, amount]) => ({ item: items[id], amount }))
      .filter(({ item, amount }) => item && amount > 0)
      .sort((item1, item2) => item1.item.id.localeCompare(item2.item.id))
);
