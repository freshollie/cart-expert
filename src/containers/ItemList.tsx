import React from "react";
import { useSelector } from "react-redux";
import { selectAllItems } from "../store/items";
import BuyableItem from "./BuyableItem";
import "./ItemList.css";

const ItemList: React.FC = () => {
  const items = useSelector(selectAllItems);

  return (
    <div className="item-list">
      {items.map((item) => (
        <BuyableItem key={item.id} item={item} />
      ))}
    </div>
  );
};

export default ItemList;
