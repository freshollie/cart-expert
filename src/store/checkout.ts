import { createSelector } from "@reduxjs/toolkit";
import { isPriceOffer, Offer, Price } from "../types";
import { selectCartItems } from "./cart";
import { selectOffers } from "./offers";

export type Discount = { offer: Offer, savings: Price };
type CheckoutDetails = {
  total: Price;
  subTotal: Price;
  discounts: Discount[];
};

export const selectCheckoutDetails = createSelector(
  [selectCartItems, selectOffers],
  (cart, offers): CheckoutDetails => {
    // find all the discounts which apply to this cart
    const discounts = offers
      .map((offer) => {
        const cartItem = cart.find(({ item }) => offer.itemId === item.id);
        const numDiscounts = Math.floor(
          (cartItem?.amount ?? 0) / offer.qualifier
        );

        if (cartItem && numDiscounts > 0) {
          return new Array(numDiscounts).fill(0).map(() => ({
            offer,
            savings: {
              amount:
                offer.qualifier * cartItem.item.price.amount -
                (isPriceOffer(offer)
                  ? offer.price.amount
                  : offer.for * cartItem.item.price.amount),
              divisor: 100,
            },
          }));
        }
        return [];
      })
      .flat();

    // create totals and subtotals
    const subTotal = {
      amount: cart.reduce(
        (acc, { amount, item }) => acc + amount * item.price.amount,
        0
      ),
      divisor: 100,
    };
    const total = {
      amount:
        subTotal.amount -
        discounts.reduce((acc, discount) => acc + discount.savings.amount, 0),
      divisor: 100,
    };

    return {
      discounts,
      subTotal,
      total,
    };
  }
);
