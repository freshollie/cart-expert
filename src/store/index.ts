import cart from "./cart";
import items from "./items";
import { configureStore } from "@reduxjs/toolkit";
import offers from "./offers";

const store = configureStore({
  reducer: {
    cart: cart.reducer,
    items: items.reducer,
    offers: offers.reducer
  },
});

export type RootState = ReturnType<typeof store.getState>;

export default store;