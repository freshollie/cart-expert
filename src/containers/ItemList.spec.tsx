import { render } from "@testing-library/react";
import React from "react";
import { Provider } from "react-redux";
import createMockStore from "redux-mock-store";
import { RootState } from "../store";
import ItemList from "./ItemList";

const exampleItems = [
  {
    id: "34583",
    name: "Beans",
    price: {
      amount: 50,
      divisor: 100,
    },
  },
  {
    id: "39845",
    name: "Coke",
    price: {
      amount: 70,
      divisor: 100,
    },
  },
  {
    id: "32844",
    name: "Oranges",
    price: {
      amount: 199,
      divisor: 100,
      perKg: true,
    },
  },
];

const mockStore = createMockStore<RootState>();

describe("ItemList", () => {
  it("should display the list of items in the store", () => {
    const { asFragment, getByText } = render(
      <Provider
        store={mockStore({
          items: exampleItems,
          offers: [],
          cart: {},
        })}
      >
        <ItemList />
      </Provider>
    );

    exampleItems.forEach((item) => expect(getByText(item.name)).toBeVisible());
    expect(asFragment()).toMatchSnapshot();
  });

  it("should display nothing when there are no items", () => {
    const { asFragment, queryByTestId } = render(
      <Provider
        store={mockStore({
          items: [],
          offers: [],
          cart: {},
        })}
      >
        <ItemList />
      </Provider>
    );

    exampleItems.forEach((item) =>
      expect(queryByTestId(item.name)).toBeFalsy()
    );
    expect(asFragment()).toMatchSnapshot();
  });
});
