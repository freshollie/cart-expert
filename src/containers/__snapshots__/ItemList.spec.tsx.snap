// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ItemList should display nothing when there are no items 1`] = `
<DocumentFragment>
  <div
    class="item-list"
  />
</DocumentFragment>
`;

exports[`ItemList should display the list of items in the store 1`] = `
<DocumentFragment>
  <div
    class="item-list"
  >
    <div
      class="item"
      data-testid="item-beans"
    >
      <h4>
        Beans
      </h4>
      <div
        data-testid="item-price"
      >
        0.50p Each
      </div>
      <div
        class="item-controls"
      >
         
        <button
          data-testid="item-remove"
          disabled=""
        >
          -
        </button>
        <div
          data-testid="cart-amount"
        >
          0
        </div>
        <button
          data-testid="item-add"
        >
          +
        </button>
      </div>
    </div>
    <div
      class="item"
      data-testid="item-coke"
    >
      <h4>
        Coke
      </h4>
      <div
        data-testid="item-price"
      >
        0.70p Each
      </div>
      <div
        class="item-controls"
      >
         
        <button
          data-testid="item-remove"
          disabled=""
        >
          -
        </button>
        <div
          data-testid="cart-amount"
        >
          0
        </div>
        <button
          data-testid="item-add"
        >
          +
        </button>
      </div>
    </div>
    <div
      class="item"
      data-testid="item-oranges"
    >
      <h4>
        Oranges
      </h4>
      <div
        data-testid="item-price"
      >
        Â£1.99 Per KG
      </div>
      <div
        class="item-controls"
      >
        <button
          data-testid="item-remove-all"
          disabled=""
        >
          Remove
        </button>
        <button>
          100g
        </button>
        <button>
          200g
        </button>
        <button>
          300g
        </button>
        <button>
          400g
        </button>
        <button>
          500g
        </button>
        <button>
          600g
        </button>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
